api/osb/reference_instance_plugin.go

```diff
@@ -0,0 +1,387 @@
+	headers.Add("Content-Type", "application/json")
+	switch method {
+	case Provision:
+		return &web.Response{
+			Body:       []byte(`{}`),
+			StatusCode: http.StatusCreated,
+			Header:     headers,
+		}, nil
+	case FetchService:
+		osbResponse, err := referencePlugin.buildOSBFetchServiceResponse(ctx, instance)
+		if err != nil {
+			return nil, err
+		}
+		marshal, err = json.Marshal(osbResponse)
+		if err != nil {
+			return nil, err
+		}
+		return &web.Response{
+			Body:       marshal,
+			StatusCode: http.StatusOK,
+			Header:     headers,
+		}, nil
+	default:
+		return &web.Response{
+			Body:       []byte(`{}`),
+			StatusCode: http.StatusOK,
+			Header:     headers,
+		}, nil
+	}
+}
+
+func (referencePlugin *referenceInstancePlugin) buildOSBFetchServiceResponse(ctx context.Context, instance *types.ServiceInstance) (osbObject, error) {
+	serviceOffering, plan, err := referencePlugin.getServiceOfferingAndPlanByPlanID(ctx, instance.ServicePlanID)
+	if err != nil {
+		return nil, util.HandleInstanceSharingError(util.ErrNotFoundInStorage, types.ServiceOfferingType.String())
+	}
+	osbResponse := osbObject{
+		"service_id":       serviceOffering.CatalogID,
+		"plan_id":          plan.CatalogID,
+		"maintenance_info": instance.MaintenanceInfo,
+		"parameters": map[string]string{
+			instance_sharing.ReferencedInstanceIDKey: instance.ReferencedInstanceID,
+		},
+	}
+	return osbResponse, nil
+}
+
+// PollBinding intercepts poll binding operation requests and check if the instance is in the platform from where the request comes
+func (referencePlugin *referenceInstancePlugin) PollBinding(req *web.Request, next web.Handler) (*web.Response, error) {
+	return referencePlugin.handleBinding(req, next)
+}
+
+func (referencePlugin *referenceInstancePlugin) validateOwnership(req *web.Request) error {
+	ctx := req.Context()
+	path := fmt.Sprintf("parameters.%s", instance_sharing.ReferencedInstanceIDKey)
+	referencedInstanceID := gjson.GetBytes(req.Body, path).String()
+	byID := query.ByField(query.EqualsOperator, "id", referencedInstanceID)
+	dbReferencedObject, err := referencePlugin.repository.Get(ctx, types.ServiceInstanceType, byID)
+	if err != nil {
+		return util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	instance := dbReferencedObject.(*types.ServiceInstance)
+	sharedInstanceTenantID := instance.Labels["tenant"][0]
+	tenantPath := fmt.Sprintf("context.%s", referencePlugin.tenantIdentifier)
+	callerTenantID := gjson.GetBytes(req.Body, tenantPath).String()
+
+	if sharedInstanceTenantID != callerTenantID {
+		log.C(ctx).Errorf("Instance owner %s is not the same as the caller %s", sharedInstanceTenantID, callerTenantID)
+		return &util.HTTPError{
+			ErrorType:   "NotFound",
+			Description: "could not find such service instance",
+			StatusCode:  http.StatusNotFound,
+		}
+	}
+	return nil
+}
+
+func (referencePlugin *referenceInstancePlugin) isReferencedShared(ctx context.Context, referencedInstanceID string) (bool, error) {
```

Comment:
you have the same function in the filter, and you wrote a utility function covering the first 5 lines, please use it..